app/services/api/open_and_special_hours_xml_to_json_parser.rb

```diff
@@ -0,0 +1,328 @@
+      end
+    end
+
+    def add_missing_to_exceptions_list(exceptions)
+      week_from = @from_hour
+      week_to = @to_hour
+      exceptions.each_with_index do |e, index|
+        if exceptions[index-1].present? && exceptions[index].present?
+          prev_pair = exceptions[index-1]
+          if e[:status] == "CLOSE" && prev_pair[:status] == "CLOSE" && Time.parse(e[:from_hour]) > Time.parse(prev_pair[:to_hour])
+            build_open_exception({
+              tmp_from: Time.parse(prev_pair[:to_hour]),
+              tmp_to: Time.parse(e[:from_hour]),
+              week_from: week_from,
+              week_to: week_to
+            })
+          end
+        end
+      end
+    end
+
+    def build_open_exception(args)
+      if (Time.parse(args[:week_from])..Time.parse(args[:week_to])).include?(args[:tmp_from]..args[:tmp_to])
+        @open_hours_and_exceptions_override[args[:tmp_from]] = {
+          from_hour: (args[:tmp_from]).strftime("%H:%M"),
+          to_hour: (args[:tmp_to]).strftime("%H:%M"),
+          status: "OPEN"
+        }
+      end
+    end
+
+    def override_from_additional_exceptions
+      tmp_exceptions = @open_hours_and_exceptions_override.sort
+      first_ex = tmp_exceptions.first
+      last_ex = tmp_exceptions.last
+
+      if first_ex.second[:from_hour] == "00:00" && last_ex.second[:to_hour] == "23:59"
+        all_open = @open_hours_and_exceptions_override.select {|h,v| v[:status] == "OPEN"}
+        if all_open.count > 0
+          # we take the first open available for the main open/close entry
+          @from_hour = all_open.first.second[:from_hour]
+          @to_hour = all_open.first.second[:to_hour]
+        end
+      end
+    end
+
+    def exception_hours(json_data)
+      json_data.select { |h,v|
+        (v[:type] == "EXCEPTION" && v[:status] == "CLOSE") || (v[:type] == "EXCEPTION" && v[:status] == "OPEN")
+      }.to_json
+    end
+
+    def all_formatted_hours
+      if all_open_exceptions.present?
+        all_open_exceptions.map {|h,v| formatted_hours(v[:from_hour],v[:to_hour])}.join(", ")
+      else
+        formatted_hours(@from_hour, @to_hour)
+      end
+    end
+
+    def all_open_hours
+      if all_open_exceptions.present?
+        all_open_exceptions.map{|h,v| {open: v[:from_hour], close: v[:to_hour]}}
+      else
+        [{open: @from_hour, close: @to_hour}]
+      end
+    end
+
+    def all_open_exceptions
+      @open_hours_and_exceptions_override.select {|h,v| v[:status] == "OPEN" }
+    end
+
+    def formatted_hours(open_time, close_time)
+      if (close_time == "00:14")
+        "#{Time.parse(open_time).strftime("%l:%M%P")} - No Closing"
+      elsif (open_time == "00:14")
+        "Closes at #{close_time}"
+      elsif (open_time == "00:00" && close_time == "23:59")
+        "Open 24 Hours"
+      elsif (open_time.eql? close_time)
+        "Closed"
+      else
+        "#{Time.parse(open_time).strftime("%l:%M%P")} -#{Time.parse(close_time).strftime("%l:%M%P")}"
```

Comment:
Just need to add a space after the `-` in this line, otherwise looks like it's good to go